File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        l = set(days)
        d = [0]
        for i in range(1,days[-1]+1):
            if i not in l:
                d.append(d[-1])
            else:
                oneday = (d[i-1] if i-1>=0 else 0)+costs[0]
                sevenday = (d[i-7] if i-7>=0 else 0)+costs[1]
                thirtyday = (d[i-30] if i-30>=0 else 0) + costs[2]
                d.append(min(oneday,sevenday,thirtyday))
        return d[-1]

            

File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.java
class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        HashSet<Integer> l = new HashSet<>();
        for(int i : days) l.add(i);

        int[] d = new int[days[days.length-1]+1];
        for(int i = 1;i<=days[days.length-1];i++){
            if(!l.contains(i)){
                d[i] = d[i-1];
            }else{
                int one = costs[0] + (i-1>=0 ? d[i-1] : 0);
                int seven = costs[1] + (i-7>=0 ? d[i-7] : 0);
                int thirty = costs[2] + (i-30>=0 ? d[i-30] : 0);
                d[i] = Math.min(one,seven);
                d[i] = Math.min(d[i],thirty);
            }
        }   
        return d[d.length-1];
    }
}